### ISSUE TITLE
'next' description in schema

### ISSUE THREAD

**Issue by pythouille** [johentsch/specs#1]
To complete the descriptions of the keys, I propose for 'next' ([here](https://github.com/johentsch/specs/blob/0efb177f12232e66274d4680b14c8b342c5cb73c/measure.schema.json#L93)) :
```
"description": "The bars that follow this one, in performance order, expressed by their 'count'. Most bars have exactly one entry in this list; those before a repeat may have two or more. The value -1 refers to the end of the piece.",
"examples": [
    [2],
    [9,1],
    [59,-1],
]
```

I could not push the change / another branch for the moment. Is there some rights to grant so I can contribute to it?

**Comment by johentsch** [comment]
Sorry, I had invited your other account. Once you accept the new invite you should be able to push.

**Comment by magiraud** [comment]
See my comments in the paper. To be non-ambiguous, next should refer to ID (or another unambiguous thing), not the number. PR #4.

**Comment by pythouille** [comment]
> To be non-ambiguous, next should refer to ID (or another unambiguous thing), not the number.

As I understand it, the values in `next` never referred to the `number` (which is ambiguous, I agree), but to `count`. At least, it is the case in the DCML '_measures.tsv' files that mentioned it originally. 

Hence, I agree with your comment in the article : `ID` is more precise than "name" (or number). But I feel that the `count` value is even better:
- It is unique, as IDs are (and in most cases, IDs will simply be a str version of the `count`)
- they allow to refer very easily to the target measure: `count` is exactly its position in the measure map when they are in list/tabular format, whereas `ID` requires more parsing to find the adequate measure in the measure map.
- finally, `count` is more compact, considering that `ID` can indicate whole links  (for example : https://dme.mozarteum.at/movi/navigator/155/001/01/20/nma, mentioned in the current schema)

In a nutshell, I feel that `ID` is more of a *global* identifier, whereas `count` is a more *local* identifier. `ID` is more general to embed logical information or reference outside the source (see link above), but `count` acts as a more concise identifier when working in a single measure map -- which is the scope of `next`, to me.

**Comment by magiraud** [comment]
I understand your point. But `count` is not musical at all. If we use count, we will have several cases where  the next of measure 16 will be something like 18 or 19.

We especially intend that the compressed format should be human-writable (and close to the musical meaning).

**Comment by johentsch** [comment]
> We especially intend that the compressed format should be human-writable (and close to the musical meaning).

That's an interesting point what would be human-writable and, true, if you write down a compressed measure map you don't have access to `count`. The "close to musical meaning" criterion does not apply to `count` at all which is purely technical and -- so far -- our next columns are also purely technical and algorithmically derived from the repeat structure and this is why, so far, the `next` column refers to `count` as @pythouille correctly writes. However I do remember a moment in our earlier discussions where we tended towards making it refer to IDs. Here's three suggestions:

* Since `ID` has strings and `count` has integers, we could allow for either type in `next`, making it clear that integers refer to counts and strings refer to IDs (could also be two separate columns but maybe we can avoid).
* Or `next` always refers to `ID` and therefore has type string.

In both cases, when you manually create the MM, the user 

* is encouraged not to create the `next` field, but 
* if they want, they should create IDs (e.g. from `number` or `name`) and use these in `next`.

Personally, I would tend towards the solution allowing for two types because it is less ambiguous and remedies the problem that I elicit [here in the PR](https://github.com/johentsch/specs/pull/4#issuecomment-1733123178). We can implement it by extending the `anyOf` clause of the JSONschema to say, either:

* `count` is present and `next` is an array of integers (or absent)
* or `ID` is present and `next is an array of strings (or absent)
* or both are present and `next` is either integers or strings.

**Comment by magiraud** [comment]
I agree with your proposal (the only danger being that "153" and 153 may then refer to different measures, but that's ok if this is well specified).

And `next` could be an array of (strings or integers), without formally enforcing that for now (because the ID/number can be given or not in the compressed MM)

**Comment by johentsch** [comment]
There is a follow-up on this [in your PR](https://github.com/johentsch/specs/pull/4#issuecomment-1733183734). Feel free to merge if you're OK with the current solution, which will close this issue.
